lib/utils/environment.ts

```diff
@@ -0,0 +1,112 @@
+limitations under the License.
+*/
+
+import * as Promise from 'bluebird';
+import { spawn } from 'child_process';
+import TypedError = require('typed-error');
+
+export interface CommandOptions {
+    /** The CWD to run the command in. */
+    cwd?: string;
+    /** How many retry attempts should occur if the command fails. */
+    retries?: number;
+}
+
+/** Used to create external command to run on the environment hosting VersionBot. */
+export interface Command {
+    /** The command to run. */
+    command: string;
+    /** The arguments to pass to the command as an array. */
+    args: string[];
+    /** Any relevant options for the command. */
+    options?: CommandOptions;
+}
+
+/**
+ * Build a new environment command object.
+ *
+ * @param   command The name of the command to execute.
+ * @param   args    An array of arguments to pass to the command.
+ * @param   options A CommandOptions structure containing options required when the command is executed.
+ * @returns         A Command object which can be passed to the RunCommand method.
+ */
+export function BuildCommand(command: string, args?: string[], options?: CommandOptions): Command {
+    let builtOptions: CommandOptions | undefined;
+    if (options) {
+        builtOptions = {
+            cwd: options.cwd,
+            retries: options.retries
+        };
+    }
+    return {
+        command,
+        args: args || [],
+        options: builtOptions
+    };
+};
+
+/**
+ * Executes an environment command.
+ * The command will be executed via the `child_process.spawn` command. This ensures that
+ * a new process is used to run the command, thus ensuring no access to the local shell.
+ * The command will be attempted based on the value of the `options.retry` property in the
+ * passed Command object. Should no retries be specified, command execution will only
+ * be attempted once.
+ *
+ * @param   params  A Command object detailing the command to execute.
+ * @returns         A Promise containing a string from `stdout`.
+ * @throws          An error containing a message from the failed command's `stderr` output.
+ */
+export function ExecuteCommand(command: Command): Promise<{}> {
+    const retries = (command.options || {}).retries || 1;
+    let retryCount = retries;
+    const callCommand = (): Promise<{}> => {
+        return new Promise((resolve, reject) => {
+            const child = spawn(command.command, command.args, command.options);
+            let stdout: string;
+            let stderr: string;
+
+            child.stdout.on('data', (data) => {
+                if (!stdout) {
+                    stdout = '';
+                }
+                stdout += data.toString();
+            });
+            child.stderr.on('data', (data) => {
+                if (!stderr) {
+                    stderr = '';
+                }
+                stderr += data.toString();
+            });
+            // If there's any stderr output, we send that instead of stdout.
```

Comment:
Anything non-zero throws, at the moment they're living with either the `stderr` or `stdout`. I'll improve this when we get round to proper error handling, too.